src/pcre2_compile.c

```diff
@@ -6541,94 +6541,63 @@ for (;; pptr++)
-          *lengthptr += 1 /* sizeof(type) in PCRE2_UCHARs */ +
-             1 /* maximum alignment. */;
-          char_lists_size >>= 1;
+          *lengthptr += 2 + LINK_SIZE;
 #else
-          *lengthptr += 1 /* sizeof(type) in PCRE2_UCHARs */;
-          /* Padding, when the size is not divisible by 4. */
-          if ((char_lists_size & 0x2) != 0)
-            char_lists_size += 2;
-          char_lists_size >>= 2;
+          *lengthptr += 1 + LINK_SIZE;
 #endif
 
-          if (OFLOW_MAX - *lengthptr < char_lists_size)
-            {
-            *errorcodeptr = ERR20;   /* Integer overflow */
-            return 0;
-            }
+          cb->char_lists_size += char_lists_size;
 
-          *lengthptr += char_lists_size;
+          char_lists_size /= sizeof(PCRE2_UCHAR);
 
-          if (*lengthptr > MAX_PATTERN_SIZE)
+          /* Storage space for character lists is included
+          in the maximum pattern size. */
+          if (*lengthptr > MAX_PATTERN_SIZE ||
+              MAX_PATTERN_SIZE - *lengthptr < char_lists_size)
             {
             *errorcodeptr = ERR20;   /* Pattern is too large */
             return 0;
             }
           }
         else
           {
-          uint8_t *char_buffer = (uint8_t*)code;
-
           PCRE2_ASSERT(cranges->char_lists_types <= XCL_TYPE_MASK);
 #if PCRE2_CODE_UNIT_WIDTH == 8
           /* Encode as high / low bytes. */
           code[0] = (uint8_t)(XCL_LIST |
             (cranges->char_lists_types >> 8));
           code[1] = (uint8_t)cranges->char_lists_types;
-          char_buffer += 2;
-
-          /* Compute alignment. */
-          if (((uintptr_t)char_buffer & 0x1) != 0)
-            {
-            code[0] |= 1u << (XCL_ALIGNMENT_SHIFT - 8);
-            char_buffer += 1;
-            }
-
-          if (((uintptr_t)char_buffer & 0x2) != (char_lists_size & 0x2))
-            {
-            code[0] |= 2u << (XCL_ALIGNMENT_SHIFT - 8);
-            char_buffer += 2;
-            }
-#elif PCRE2_CODE_UNIT_WIDTH == 16
-          code[0] = (PCRE2_UCHAR)(XCL_LIST | cranges->char_lists_types);
-          char_buffer += 2;
-
-          /* Compute alignment. */
-          if (((uintptr_t)char_buffer & 0x2) != (char_lists_size & 0x2))
-            {
-            code[0] |= 2u << XCL_ALIGNMENT_SHIFT;
-            char_buffer += 2;
-            }
+          code += 2;
 #else
-          code[0] = (PCRE2_UCHAR)(XCL_LIST | cranges->char_lists_types);
-          char_buffer += 4;
-
-          /* Padding. */
-          if ((char_lists_size & 0x2) != 0)
-            {
-            code[0] |= 2u << XCL_ALIGNMENT_SHIFT;
-            char_buffer += 2;
-            }
+          *code++ = (PCRE2_UCHAR)(XCL_LIST | cranges->char_lists_types);
 #endif
-          memcpy(char_buffer,
+
+          cb->char_lists_size += char_lists_size;
+
+          memcpy((uint8_t*)cb->start_code - cb->char_lists_size,
             (uint8_t*)(cranges + 1) + cranges->char_lists_start,
             char_lists_size);
```

Comment:
You could memset the padding bytes to 0xff or something. It appears you're just leaving whatever junk is in there from malloc?

If you do initialize... then you'll get predictable runtime behaviour (if something reads from it by mistake). If you don't initialize... then you'll get nice valgrind warnings.

Reply:
Make tools happy. Ok, I will check.

Reply:
The best-of-both would be a memset to 0xff and also an `ifdef SUPPORT_VALGRIND ... VALGRIND_MAKE_MEM_NOACCESS`.

Reply:
I prefer doing only the VALGRIND_MAKE_MEM_NOACCESS, so any mistaken reads can be detected, after all those valgrind warnings are a feature!, the memset is just a workaround IMHO.

Reply:
I have added a `((uint16_t*)data)[-1] = 0xffff;` to set all bits. Btw valgrind only complains if you check the memory. Just copying it is not a problem.

Reply:
Just to check I understand - the data is 16-bit and/or 32-bit packed values, but we want it to be 32-bit aligned. So the `data` variable is always 16-bit aligned, and if there is padding, it will be exactly 2 bytes. So the cast `(uint16_t*)data` is doing a correctly-aligned write of the two padding bytes.

That all looks great, thanks!

Reply:
Yes. The character list data is a stream of 16 bit values followed by a stream of 32 bit values. Both stream can be empty. Regardless, the final data is considered to be a stream of 32 bit values, and there might be an unused 16 bit value in the beginning sometimes.